### PR TITLE
Dispatcher refactoring

### PR DESCRIPTION
**Pull request by ssd04**
Dispatcher refactoring:
- remove `graphql` implementation
- refactor websockets setup by decoupling components
     - add mocks and unit tests
   
Added `cmd/notifier` and moved main.go and `config.toml` here, in order to align cmd setup with the other repos.

Updated config file:
- redis connection type
- remove `dispatcherType` and `hubType` since they are not being used

### CODE REVIEW

File: dispatcher/hub/commonHub.go [hunk 1]
```diff
@@ -13,10 +14,16 @@ import (
 
 var log = logger.GetOrCreate("hub")
 
+// ArgsCommonHub defines the arguments needed for common hub  creation
```
Comment (bogdan-rosianu) [inline dispatcher/hub/commonHub.go:17]: 2 spaces between `hub` and `creation`

File: dispatcher/ws/wsHandler.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	"github.com/ElrondNetwork/notifier-go/dispatcher"
+)
+
+// ArgsWebSocketHandler defines the argument needed to create a websocketHandler
+type ArgsWebSocketHandler struct {
+	Hub      dispatcher.Hub
+	Upgrader dispatcher.WSUpgrader
+}
+
+type websocketHandler struct {
```
Comment (bogdan-rosianu) [inline dispatcher/ws/wsHandler.go:16]: can we rename it? we usually name interfaces with a `handler` suffix

File: facade/notifierFacade.go [hunk 3]
```diff
@@ -60,14 +62,9 @@ func (nf *notifierFacade) HandleFinalizedEvents(events data.FinalizedBlock) {
...
 
-// GetDispatchType will provide dispatch type based on config
-func (nf *notifierFacade) GetDispatchType() string {
-	return nf.config.DispatchType
-}
-
-// GetHub will return the hub component
-func (nf *notifierFacade) GetHub() dispatcher.Hub {
-	return nf.hub
+// Server will handle a websocket request
```
Comment (bogdan-rosianu) [inline facade/notifierFacade.go:65]: wrong comment

File: factory/hubFactory.go [hunk 4]
```diff
@@ -1,62 +1,34 @@
...
-	}
-
-	return tryMakeCustomHubForID(hubConfig[1])
-}
-
-func tryMakeCustomHubForID(id string) (dispatcher.Hub, error) {
-	if makeHubFunc, ok := availableHubDelegates[id]; ok {
-		customHub := makeHubFunc()
-		return customHub, nil
+	commonHub, err := hub.NewCommonHub(args)
```
Comment (bogdan-rosianu) [inline factory/hubFactory.go:28]: could have been `return hub.NewCommonHub(args)`

File: factory/wsFactory.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+	upgrader, err := ws.NewWSUpgraderWrapper(readBufferSize, writeBufferSize)
+	if err != nil {
+		return nil, err
+	}
+
+	args := ws.ArgsWebSocketHandler{
+		Hub:      hub,
+		Upgrader: upgrader,
+	}
+	wsHandler, err := ws.NewWebSocketHandler(args)
```
Comment (bogdan-rosianu) [inline factory/wsFactory.go:37]: could have been `return ws.NewWebSocketHandler(args)`

File: mocks/wsHandlerStub.go [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import "net/http"
+
+// WSHandlerStub implements WSHandler interface
+type WSHandlerStub struct {
+	ServeHTTPCalled func(w http.ResponseWriter, r *http.Request)
+}
+
+// ServeHTTP -
+func (whs *WSHandlerStub) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+}
```
Comment (bogdan-rosianu) [inline mocks/wsHandlerStub.go:12]: forgot to call `ServeHTTPCalled` on this function

File: process/eventsHandler.go [hunk 7]
```diff
@@ -67,12 +67,17 @@ func checkArgs(args ArgsEventsHandler) error {
 
 // HandlePushEvents will handle push events received from observer
 func (eh *eventsHandler) HandlePushEvents(events data.BlockEvents) {
+	if events.Hash == "" {
```
Comment (bogdan-rosianu) [inline process/eventsHandler.go:70]: might worth setting these logs as `WARN` ? I think it shouldn't happen

File: dispatcher/ws/wsHandler.go (hunk 2 above)
Comment (ssd04) [reply dispatcher/ws/wsHandler.go]: renamed with `processor` suffix

File: mocks/wsHandlerStub.go (hunk 6 above)
Comment (ssd04) [reply mocks/wsHandlerStub.go:15]: right :+1: thanks

File: process/eventsHandler.go (hunk 7 above)
Comment (ssd04) [reply process/eventsHandler.go:70]: set the logs to WARN

**Review by bogdan-rosianu — DISMISSED** [review]

File: dispatcher/hub/commonHub.go [hunk 8]
```diff
@@ -28,19 +35,35 @@ type commonHub struct {
...
+	if err != nil {
+		return nil, err
+	}
+
 	return &commonHub{
 		rwMut:              sync.RWMutex{},
-		filter:             eventFilter,
-		subscriptionMapper: dispatcher.NewSubscriptionMapper(),
+		filter:             args.Filter,
+		subscriptionMapper: args.SubscriptionMapper,
```
Comment (iulianpascalau) [inline dispatcher/hub/commonHub.go:47]: :+1:

File: dispatcher/ws/wsDispatcher_test.go [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+	"github.com/ElrondNetwork/notifier-go/data"
+	"github.com/ElrondNetwork/notifier-go/dispatcher/ws"
+	"github.com/ElrondNetwork/notifier-go/mocks"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type testWriter struct{}
+
+func (tw *testWriter) Write(p []byte) (n int, err error) {
```
Comment (iulianpascalau) [inline dispatcher/ws/wsDispatcher_test.go:17]: missing comments?

**Review by iulianpascalau — APPROVED** [review]

File: dispatcher/ws/wsDispatcher_test.go (hunk 9 above)
Comment (ssd04) [reply dispatcher/ws/wsDispatcher_test.go:19]: added comments

**Review by iulianpascalau — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Pull request merged**